ADR: initial design (#1)
filecoin-station/zinnia: docs/architecture-decision-records/2023-01-initial-architecture.md

- It's a VC-funded project with goals that may not align with our goals.

- It provides additional APIs that we may not want to expose to modules. As a
  result, we are likely to spend extra effort to disable or sandbox these APIs.

- User-facing APIs like streams are modelled for TypeScript. Exposing Rust
  crates for Rust/WASM consumption will require unnecessary overhead, switching
  from Rust to TypeScript idioms and then back to Rust again.

### Wasmtime

_https://wasmtime.dev/_

Wasmtime is a fast and secure runtime for WebAssembly, it's a
[Bytecode Alliance](https://bytecodealliance.org) project.

Benefits:

- A community-driven project not depending on any single company

- It implements future proposals to WebAssembly and helps drive innovation in
  the WebAssembly space, e.g. WebAssembly Interface Types and Component Model.

- The runtime is not opinionated about networking & filesystem APIs. It's
  possible to add WASI for filesystem access, but it's also easy to use
  proprietary host APIs instead.

- We have much more control over the I/O layer. It should be easier to implement
  resource usage limiting.

- It implements gas metering & limiting.

> - FMV is built on top of Wasmtime too. This can create synergies between our

Reviewer: FVM

Reviewer 2: ```suggestion
- FVM is built on top of Wasmtime too. This can create synergies between our
```

  teams.

- It supports async host functions.

- Comprehensive documentation for embedders.

Issues:

- Apple M1 architecture has Tier 3 support only (not production ready).

- It's not clear whether functions exported by WASM modules can be async. This
  may require custom glue code to be written.

- There is no stable API for HTTP/HTTPS. WASI does not support opening new
  network connections yet, and API for crypto primitives required by TLS is not
  specified yet.

- Supports WASM only, it cannot run JavaScript/TypeScript code.

### WasmEdge

_https://wasmedge.org/_

WasmEdge aims to bring the cloud-native and serverless application paradigms to
Edge Computing. It's in many aspects similar to Wasmtime. Unfortunately, it's a
C/C++ project, which makes it difficult to integrate into Rust projects,
especially on Apple Silicon, where binaries must be signed.

### Wasmer

_https://wasmer.io_

Wasmer is similar to Wasmtime.

Benefits:

- WebAssembly package manager offering a catalogue of WASM modules

- Supports WASI (filesystem)

- Popular in Web3 space (ChainSafe, Fluence, Hyperledger, and more)

Issues:

> - No support for an asynchronous programming model & non-blocking I/A

Reviewer 2: ```suggestion
- No support for an asynchronous programming model & non-blocking I/O
```

Runtime.**

While Deno would allow us to start faster, by providing Fetch API and Filesystem
access with built-in sandboxing, I feel it would make it much harder for us to
build past the built-in features (add IPFS & libp2p, implement resource usage
limiting). Our requirements are unlikely to align with Denoland's vision for
Deno, so we may have a hard time getting our changes accepted.

We are going to build our Runtime incrementally. Early Module authors will
likely need to contribute new features to our Runtime to enable their modules to
be built. This will be easier if the interface between modules and the runtime
is Rust-native. If we built on top of Deno with wasm-bindgen, js-sys and
web-sys, then module authors would need to understand both Rust and JavaScript.

That's why I believe it's better to choose a more bare-bone WASM engine.

From the different WASM engine options out there, it seems that Wasmtime is best
aligned with ProtocolLab's vision of driving technical breakthroughs in the
open. The project is community-driven and contributes to other WASM initiatives.

On the technical side, it is the only engine that supports the asynchronous

> programming model (AFAICT). It also has the best documentation.

Reviewer 2: ```suggestion
programming model (AFAICT). It also has the most extensive documentation.
```

I try to avoid words like "best", because they can mean different things to different people, and replace them with something more semantic. Does this work for you?

1. Sandboxing
2. Resource usage limit
3. Modules can be written in any language, composed from multiple heterogenous components
   (e.g. nginx reverse proxy + Node.js app)

However, we ruled out virtualization-based solutions for the following reasons:

  1. Entry-level computers may not support CPU instructions required for virtualization.
  2. Often, virtualization must be enabled in BIOS settings.
  3. Windows Home edition does not support Microsoft's virtualization framework Hyper-V.
  4. Sharing the same IPFS & libp2p node between multiple container instances is difficult.

### Deno

_https://deno.land_

Deno offers JS & WASM runtime with filesystem and networking APIs, including capabilities-based
sanboxing.

Benefits:

- Full support for both JavaScript/TypeScript and WebAssembly
- Filesystem and network access is sandboxed out of the box
- The engine and platform APIs are designed for asynchrony and non-blocking I/O
- WebAPI-compliant Fetch API implementation
- WebAPIs can be easily called from Rust via [wasm-bindgen](https://crates.io/crates/wasm-bindgen) and [web-sys](https://crates.io/crates/web-sys) crates.

Issues:

- Missing documentation for embedders

> - Static JS imports are not sandboxed. Add this line to your TS script `import * as google from 'https://google.com/'` - it will execute even if network access is not allowed.

Reviewer 2: ouch! Can you run `deno bundle` to create a script that will have this `import` already inlined? I'm quite sure afterwards there will be no more network calls.

Author: Yeah, if we wanted to use Deno, then we should require modules to provide us a single file with the bundle of all source code.

Then we would have to implement parse that code to ensure it does not contain any static imports.

I think that should be easily doable 👍🏻

Reviewer 2: If we use deno, we could also automate the processes of using `deno build` as part of importing/updating the module
